File: examples/canonical-analytics/python/src/analytics_workbench/charts/service_bar.py
import plotly.graph_objects as go
from ..analytics.services import ServiceErrorRates
from datetime import datetime

def build_service_bar_figure(agg: ServiceErrorRates) -> dict:
    fig = go.Figure()
    
    total_count = sum(agg.total_events) if agg.total_events else 0
    
    if not agg.services:
        fig.add_annotation(text="No data available", showarrow=False, font=dict(size=20))
        fig.update_layout(title_text="Service Error Rates", xaxis_title_text="Error Rate", yaxis_title_text="Service")
    else:
        # We need horizontal sorted bar, sorted descending by error rate
        # Plotly draws horizontal bars from bottom up, so to show highest at top we reverse the lists
        services_rev = agg.services[::-1]
        rates_rev = agg.error_rates[::-1]
        
        fig.add_trace(go.Bar(
            x=rates_rev,
            y=services_rev,
            orientation="h",
            name="Error Rate",
            hovertemplate="Service: %{y}<br>Error Rate: %{x:.2%}<extra></extra>"
        ))
        
        fig.update_layout(
            title_text="Service Error Rates",
            xaxis_title_text="Error Rate",
            yaxis_title_text="Service",
            xaxis_tickformat=".0%"
        )
        
    fig_dict = fig.to_dict()
    fig_dict["meta"] = {
        "total_count": total_count,
        "filters_applied": [],
        "generated_at": datetime.utcnow().isoformat()
    }
    return fig_dict

File: examples/canonical-analytics/rust/src/charts/service_bar.rs
use super::to_figure_json;
use crate::analytics::ServiceErrorRates;
use plotly::common::{Orientation, Title};
use plotly::{Bar, Plot};
use serde_json::Value;

pub fn build_service_bar_figure(agg: &ServiceErrorRates) -> Value {
    let mut plot = Plot::new();

    if agg.services.is_empty() {
        let trace = Bar::new(Vec::<f64>::new(), Vec::<String>::new()).orientation(Orientation::Horizontal);
        plot.add_trace(trace);
        let layout = plotly::Layout::new().title(Title::with_text("No data for selected filters"));
        plot.set_layout(layout);
        return to_figure_json(&plot, 0, true);
    }

    // Sort ascending for Plotly horizontal bar so top is at the top
    let mut srvs = agg.services.clone();
    srvs.reverse();
    let mut rates = agg.error_rates.clone();
    rates.reverse();
    
    let total_count: i64 = agg.total_events.iter().sum();

    let trace = Bar::new(rates, srvs)
        .orientation(Orientation::Horizontal)
        .hover_template("Service: %{y}<br>Error Rate: %{x:.2%}<extra></extra>");
    
    plot.add_trace(trace);

    let layout = plotly::Layout::new()
        .title(Title::with_text("Service Error Rates"))
        .x_axis(plotly::layout::Axis::new().title(Title::with_text("Error Rate")))
        .y_axis(plotly::layout::Axis::new().title(Title::with_text("Service")));
    plot.set_layout(layout);

    to_figure_json(&plot, total_count, true)
}
